### PR TITLE
Tasmanian mismatch

### PR DESCRIPTION
**Pull request by aerijman**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [x] - This PR updates an existing tool or tool collection
* [x] - This PR does something else (explain below)

### CODE REVIEW

**Comment by aerijman** [comment]
Please, review this wrapper of a tool that assesses the damage of samples of DNA. Summarized in a short sentence, It counts mismatches per position and produces a table and scatter plots in html that is embedded in galaxy. 

Thank you!

File: tools/tasmanian-mismatch/Tasmanian.xml [hunk 1]
```diff
@@ -0,0 +1,152 @@
+<tool id="Tasmanian" name="Analysis of artifacts with Tasmanian" version="0.1.2" python_template_version="3.7">
```
Comment (bgruening) [inline tools/tasmanian-mismatch/Tasmanian.xml:1]: the tool-id should just contain small letters

File: tools/tasmanian-mismatch/Tasmanian.xml [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  <requirements>
+    <requirement type="package" version="3.7.4">python</requirement>
+    <requirement type="package" version="0.1.3">tasmanian-mismatch</requirement>
```
Comment (bgruening) [inline tools/tasmanian-mismatch/Tasmanian.xml:4]: maybe use the same version of the tool as this version? `version="0.1.3"`

File: tools/tasmanian-mismatch/Tasmanian.xml [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    <requirement type="package" version="1.17.2">numpy</requirement>
+    <requirement type="package" version="1.3.1">scipy</requirement>
+    <requirement type="package" version="4.1.1">plotly</requirement>-->
+    <requirement type="package" version="1.9">samtools</requirement>
+  </requirements>
+
+  <command detect_errors="exit_code">
+    <![CDATA[
+   
+     ##conda create -n tasmanian python=3 --yes;
```
Comment (bgruening) [inline tools/tasmanian-mismatch/Tasmanian.xml:15]: can you please clean this up and remove all lines that are not needed?

File: tools/tasmanian-mismatch/Tasmanian.xml [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+     ##conda activate tasmanian;
+     ##conda install -c bioconda tasmanian-mismatch --yes;
+
+     #set $reference_fasta_filename = "localref.fa"
+
+     #if str( $reference_source.reference_source_selector ) == "history":
+        ln -s '${reference_source.ref_file}' '${reference_fasta_filename}' &&
+     #else:
+        #set $reference_fasta_filename = str( $reference_source.ref_file.fields.path )
```
Comment (bgruening) [inline tools/tasmanian-mismatch/Tasmanian.xml:24]: that should also be single-quoted

File: tools/tasmanian-mismatch/Tasmanian.xml [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+     #end if
+    
+     #if str( $bed_file_source.bed_file_selector ) == "yes":
+            #set $bed_filename = str ( $bed_file_source.bed_file )
+            samtools view '$input1' | run_intersections -b  '$bed_filename' | run_tasmanian -r '$reference_fasta_filename' > '$output1'
+     #else        
+            samtools view '$input1' | run_tasmanian -r '$reference_fasta_filename' > '$output1'
+     #end if  
+
+     ##samtools view '$input1' | run_intersections -b '$__tool_directory__/bed_file_Galaxy' | run_tasmanian -r '$reference_fasta_filename' > '$output1'
```
Comment (bgruening) [inline tools/tasmanian-mismatch/Tasmanian.xml:34]: can this be removed?

File: tools/tasmanian-mismatch/Tasmanian.xml [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+     ##samtools view '$input1' | python3 '$__tool_directory__/run_intersections.py' -b '$__tool_directory__/bed_file_Galaxy' | python3 '$__tool_directory__/run_tasmanian.py' -r '$reference_fasta_filename' > '$output1'
+
+  ]]></command>
+  <inputs>
+    <!-- Bam alignment file -->
+    <param type="data" name="input1" label="bam alignemnt file" format="bam" />
```
Comment (bgruening) [inline tools/tasmanian-mismatch/Tasmanian.xml:40]: please improve a little bit the label and the help section of your parameters

File: tools/tasmanian-mismatch/Tasmanian.xml [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    <conditional name="bed_file_source">
+      <param name="bed_file_selector" type="select" label="bed_file">
+        <option value="no">No don't use bed file </option>
+        <option value="yes">Yes use bed file</option>
+      </param>
+      <when value="yes">
+        <param name="bed_file" type="data" format="bed" label="Use the following dataset" />
+      </when> 
+      <when value="no">
+        <param name="bed_file" value="0" type="integer" label="Not defining regions in a bed file" />
```
Comment (bgruening) [inline tools/tasmanian-mismatch/Tasmanian.xml:71]: what does this parameter do?

Do you want to create an optional `bed_file` parameter? Then use optional="True" and remove the entire conditional.

File: tools/tasmanian-mismatch/Tasmanian.xml [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+      </when>
+    </conditional>
+
+    <!-- Additional parameters -->
+    <param name="confidence" label="Confidence" type="integer" value="20" help="Number of bases in boundary region, from 0 to length of the read (read help below). Default=20"/>
```
Comment (bgruening) [inline tools/tasmanian-mismatch/Tasmanian.xml:76]: can you add min and/or max to those parameters?

File: tools/tasmanian-mismatch/Tasmanian.xml [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+  <tests>
+    <!-- test when reference from history -->
+    <test>
+      <param name="input1" value="test2.bam" ftype="bam"/>
+      <param name="reference_source_selector" value="history"/>
+      <param name="ref_file" value="small_region.fa"/>
+      <output name="output1" file="test2.output" lines_diff="4"/>
+    </test>
+    <!-- test when reference from cached
```
Comment (bgruening) [inline tools/tasmanian-mismatch/Tasmanian.xml:94]: There are a few examples in this repo with tests for reference datasets.

File: tools/tasmanian-mismatch/Tasmanian.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (aerijman) [reply tools/tasmanian-mismatch/Tasmanian.xml:40]: If using a bed file Sam does not need to be sorted but without a bed file it should be sorted 

Should I check and throw an error or should I sort it if not sorted?

Sam is also possible, if sorted... Should I include the option?

[the remark this answers is not in the document]
Comment (bernt-matthias) [reply tools/tasmanian-mismatch/Tasmanian.xml:40]: Sounds like a good idea. In cheetah you can check the data type with 

```
#if $data.ext in ["sam", "unsorted.bam"]
   ...
#end if 
```

Here you may insert a `samtools sort ... | `

There are also bam data types that are sorted in other ways (https://github.com/galaxyproject/galaxy/blob/aaa8a2f7a82b1745a8faa8475507e9d6062198dd/lib/galaxy/config/sample/datatypes_conf.xml.sample#L26 and the following lines). Maybe the query sorted bam types or cram can be covered as well.

Comment (aerijman) [reply tools/tasmanian-mismatch/Tasmanian.xml:40]: Please, let me know if it needs further improvement.

File: tools/tasmanian-mismatch/Tasmanian.xml (hunk 7 above)
Comment (aerijman) [reply tools/tasmanian-mismatch/Tasmanian.xml:71]: Thanks a lot!! Now it looks much better!

File: tools/tasmanian-mismatch/Tasmanian.xml (hunk 6 above)
Comment (bernt-matthias) [reply tools/tasmanian-mismatch/Tasmanian.xml:40]: Seems that you did not push your changes yet?

File: tools/tasmanian-mismatch/Tasmanian.xml [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+     #if str( $reference_source.reference_source_selector ) == "history":
+        ln -s '${reference_source.ref_file}' '${reference_fasta_filename}' &&
+     #else:
+        #set $reference_fasta_filename = str( $reference_source.ref_file.fields.path )
+     #end if
+    
+     #if str( $bed_file_source.bed_file_selector ) == "yes":
+            #set $bed_filename = str ( $bed_file_source.bed_file )
+            samtools view '$input1' | run_intersections -b  '$bed_filename' | run_tasmanian -r '$reference_fasta_filename' > '$output1'
```
[the remark this answers is not in the document]
Comment (aerijman) [reply tools/tasmanian-mismatch/Tasmanian.xml:29]: Thank you, it reads a lot better now!

File: tools/tasmanian-mismatch/Tasmanian.xml [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+    <!-- reference genome upload -->
+    <conditional name="reference_source">
+      <param name="reference_source_selector" type="select" label="Load reference genome from">
+        <option value="cached">Local cache</option>
+        <option value="history">History</option>
+      </param>
+      <when value="cached">
+        <param name="ref_file" type="select" label="Use dictionary from the list" help="Select genome from the list">
+          <options from_data_table="tasmanian_indexes">
+            <filter type="sort_by" column="2" />
```
[the remark this answers is not in the document]
Comment (aerijman) [reply tools/tasmanian-mismatch/Tasmanian.xml:39]: By column name ("_dbkey_")?  
 
I copied this from other tools and it seems to work in them. I am missing something that might not be clear [HERE](https://planemo.readthedocs.io/en/latest/writing_how_do_i.html#test-index-loc-data)?

File: tools/tasmanian-mismatch/Tasmanian.xml [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+    <!-- Additional parameters -->
+    <param name="confidence" label="Confidence" type="integer" value="20" help="Number of bases in boundary region, from 0 to length of the read (read help below). Default=20"/> 
+    <param name="softlips" label="Softclips" type="integer" value="0" help="How sofclips whould be treated. Values include 0,1 or 2 (read the help below). Default=0"/>
+    <param name="mapquality" label="Map quality" type="integer" value="20" help="Exclude bases with lower mapQ than this number. Default=20"/>
+  </inputs>
+
+  <outputs>
+    <data name="output1" format="txt" /> <!--output" />-->
+    <data format="html" name="html_file" from_work_dir="Tasmanian_artifact_report.html" label="Tasmanian results" />
```
[the remark this answers is not in the document]
Comment (aerijman) [reply tools/tasmanian-mismatch/Tasmanian.xml:65]: prefix=done.

The HTML simplifies considerably the analysis and I believe is worth keeping it here. 
I am open to take it out if you think it could be a headache for anyone unable to manage the whitelist.

File: tools/tasmanian-mismatch/Tasmanian.xml [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+    <test>
+      <param name="input1" value="test2.bam" ftype="bam"/>
+      <param name="reference_source_selector" value="history"/>
+      <param name="ref_file" value="small_region.fa"/>
+      <output name="output1" file="test2.output" lines_diff="4"/>
+    </test>
+    <!-- test when reference from cached
+    <test>
+      <param name="input1" value="test2.bam" ftype="bam" dbkey="chrI"/>
+      <param name="reference_source_selector" value="cached"/>
```
[the remark this answers is not in the document]
Comment (aerijman) [reply tools/tasmanian-mismatch/Tasmanian.xml:79]: I commented to @bgruening that I don't know what I am missing here. I spend considerably amount of time trying to elucidate why just doing [THIS](https://planemo.readthedocs.io/en/latest/writing_how_do_i.html#test-index-loc-data) did not work and have been unsuccessful so far. I will greatly appreciate if you (or other) could give me a clue or teach me what's going on. Thanks!

File: tools/tasmanian-mismatch/Tasmanian.xml [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+        <param name="bed_file" type="data" format="bed" label="Use the following dataset" />
+      </when> 
+      <when value="no">
+        <param name="bed_file" value="0" type="integer" label="Not defining regions in a bed file" />
+      </when>
+    </conditional>
+
+    <!-- Additional parameters -->
+    <param name="confidence" label="Confidence" type="integer" value="20" help="Number of bases in boundary region, from 0 to length of the read (read help below). Default=20"/> 
+    <param name="softlips" label="Softclips" type="integer" value="0" help="How sofclips whould be treated. Values include 0,1 or 2 (read the help below). Default=0"/>
```
[the remark this answers is not in the document]
Comment (aerijman) [reply tools/tasmanian-mismatch/Tasmanian.xml]: `Select` now here!

File: tools/tasmanian-mismatch/Tasmanian.xml (hunk 11 above)
[the remark this answers is not in the document]
Comment (bernt-matthias) [reply tools/tasmanian-mismatch/Tasmanian.xml:39]: Yep .. given that dbkey is the name of the 2nd column. Using column numbers is used quite often, but the column name is easier to read.

File: tools/tasmanian-mismatch/Tasmanian.xml (hunk 12 above)
[the remark this answers is not in the document]
Comment (bernt-matthias) [reply tools/tasmanian-mismatch/Tasmanian.xml:65]: For sure. My suggestion was to ass an additional bool input param (e.g. `keep_html`) and a filter for this output that depends on the bool param. Then the user has the choice to remove the output. 

You may default the boolean to true if you like.

File: tools/picard/tool_data_table_conf.xml.test [hunk 15]
```diff
@@ -1,7 +1,8 @@
+<!-- Use the file tool_data_table_conf.xml.oldlocstyle if you don't want to update your loc files as changed in revision 4550:535d276c92bc-->
 <tables>
     <!-- Location of Picard dict file and other files -->
     <table name="picard_indexes" comment_char="#">
         <columns>value, dbkey, name, path</columns>
-        <file path="${__HERE__}/test-data/cached_locally/picard_indexes.loc" />
```
Comment (bernt-matthias) [inline tools/picard/tool_data_table_conf.xml.test:5]: Please don't change the tool_data_table_conf.xml.test in picard, even if you use the same data table.

- Copy the tool_data_table_conf.xml.test and tool_data_table_conf.xml files from picard (and maybe adapt as needed)
- Keep the `${__HERE__}/test-data/cached_locally/picard_indexes.loc` but change it such that it points to an existing file
- In the loc file in test-data you need to refer to an existing test data file .. I remember that you need to use `${__HERE__}` as well to refer to the actual test data

I guess you can find this is one of the tools.

File: tools/tasmanian-mismatch/Tasmanian.xml (hunk 13 above)
[the remark this answers is not in the document]
Comment (bernt-matthias) [reply tools/tasmanian-mismatch/Tasmanian.xml:79]: Yeah this is a bit tricky. You can just uncomment the test and we will help debugging if needed. I added a few more comments on this here: https://github.com/galaxyproject/tools-iuc/pull/2971/files#r410751577

File: tools/tasmanian-mismatch/Tasmanian.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (aerijman) [reply tools/tasmanian-mismatch/Tasmanian.xml]: I left it as is:  
Please, apologize me if I got it completely wrong.  
  

- Sam can still be input for samtools view. 
- It looks to me that data.ext checks the suffix but different users have different habits naming these files. 

I will make sure that the `Error` is clear in *Tasmanian-mismatch* if the input is not sorted.

File: tools/tasmanian-mismatch/Tasmanian.xml (hunk 12 above)
[the remark this answers is not in the document]
Comment (aerijman) [reply tools/tasmanian-mismatch/Tasmanian.xml]: Cool, done! I hope that I did it the right way.

File: tools/picard/tool_data_table_conf.xml.test (hunk 15 above)
Comment (aerijman) [reply tools/picard/tool_data_table_conf.xml.test:5]: ln ../tools-iuc/tools/tasmanian-mismatch:

```
...
tool_data_table_conf.xml.test
tool_data_table_conf.xml.sample
```
---

tool_data_table_conf.xml.test
```
<tables>
    <!-- Location of Picard dict file and other files -->
    <table name="tasmanian_indexes" comment_char="#">
        <columns>value, dbkey, name, path</columns>
        <file path="${__HERE__}/test-data/cached_locally/tasmanian_indexes.loc" />
    </table>
</tables>
```
tool_data_table_conf.xml.sample
```
<tables>
    <!-- Location of Picard dict file and other files -->
    <table name="tasmanian_indexes" comment_char="#">
        <columns>value, dbkey, name, path</columns>
        <file path="tool-data/tasmanian_indexes.loc" />
    </table>
</tables>
```
---  
/test-data/cached_locally/tasmanian_indexes.loc
`chrI	chrI	ChrI	${__HERE__}/small_region.fa`

I used dbkey="chrI" although I saw that other Picard tools should have used dbkeys different than hg38 but using hgh38 it works. Therefore I also tried with hg38 but without success.

File: tools/tasmanian-mismatch/Tasmanian.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (bernt-matthias) [reply tools/tasmanian-mismatch/Tasmanian.xml]: > It looks to me that data.ext checks the suffix but different users have different habits naming these files.

This is the data type set in Galaxy. It does not matter which extensions users choose.

**Comment by bernt-matthias** [comment]
Currently for some reason only the picard tools are tests. Seems that the reason is that your master branch and the IUC master branch were quite out of sync when you created the branch of this feature request. 

To me it seems like you should:

- drop https://github.com/galaxyproject/tools-iuc/pull/2971/commits/352dc9e3be3cb2562d69db6d795bb4cec0c4abd5 and https://github.com/galaxyproject/tools-iuc/pull/2971/commits/3de5be64e970d7d3666b5517d067fcb7b6847876
- sync your master and the IUC  master 
- rebase this branch
- revert the changes to `tools/picard/tool_data_table_conf.xml.test`

**Comment by bgruening** [comment]
@aerijman if you have problems with that git stuff, please let us know and we can give you a hand.

**Comment by aerijman** [comment]
@bgruening and @bernt-matthias Thank you very much for your help!!
I presumably synced my master to upstream and rebased Tasmanian-mismatch.
I guess I am not being assertive in committing the changes. If this is the case, I accept your help!!

**Comment by bgruening** [comment]
> Should I make a new a pull-request?

If you are not sure. Make a new PR and we can see :)

Thanks @aerijman!

File: tools/tasmanian_mismatch/.shed.yml [hunk 16]
```diff
@@ -0,0 +1,7 @@
+categories: [Sequence Analysis]
+description: Analysis of positional mismatches
+long_description: | 
+A tool to count and visualize the distribution of positional mismatches on reads of High throughput DNA sequencing experiments
```
Comment (bernt-matthias) [inline tools/tasmanian_mismatch/.shed.yml:4]: I think you need to indent this?

File: tools/tasmanian_mismatch/.shed.yml [hunk 17]
```diff
@@ -0,0 +1,7 @@
+categories: [Sequence Analysis]
```
Comment (bernt-matthias) [inline tools/tasmanian_mismatch/.shed.yml:1]:
```
categories: 
- Sequence Analysis
```

File: tools/picard/tool_data_table_conf.xml.test [hunk 18]
```diff
@@ -1,7 +1,8 @@
+<!-- Use the file tool_data_table_conf.xml.oldlocstyle if you don't want to update your loc files as changed in revision 4550:535d276c92bc-->
 <tables>
```
Comment (bernt-matthias) [inline tools/picard/tool_data_table_conf.xml.test:2]: The change to `tools/picard/tool_data_table_conf.xml.test` needs to be reverted. Also, please enable the test with cached data. :) 

Great work so far.

Comment (aerijman) [reply tools/picard/tool_data_table_conf.xml.test]: @bernt-matthias thank you!

I am puzzled now because planemo is currently not working after changing Picard file with error:  

>  Cannot locate xUnit report [/private/var/folders/23/72zvgxhx7yxd_q5kyb_vdt9w0000gq/T/tmpA3DNnu/xunit.xml]

I currently changed ` tools/picard/tool_data_table_conf.xml.test`. I can't track when I changed that file. 


The test for cache is uncommented now.

**Comment by aerijman** [comment]
Hi, I don't understand what is failing. Would you mind guiding me on that?

**Comment by bgruening** [comment]
> tool_data_table_conf.xml.test

is not well-formatted XML file.

Please reset all changes to tools/picard/tool_data_table_conf.xml.test

File: tools/picard/tool_data_table_conf.xml.test (hunk 15 above)
Comment (bgruening) [reply tools/picard/tool_data_table_conf.xml.test:5]: @aerijman you really should not touch this file at all.

Can you do a "git checkout master tools/picard/tool_data_table_conf.xml.test" to restore the original, please.

**Comment by aerijman** [comment]
Hi, 

1 - I hope it's ready to merge!
2 - .shed.yml linted correctly. Should I create empty tool_dependencies.xml and repository_dependencies.xml files?  
Thank you!

**Comment by nsoranzo** [comment]
tool_dependencies.xml and repository_dependencies.xml are mostly unnecessary nowadays, relics from pre-Conda says.

**Comment by bgruening** [comment]
@aerijman please do not manipulate files in another folder. If you think you need to fix picard, can you do this in a different PR?

**Comment by aerijman** [comment]
@bgruening Sure, I will follow your suggestions! 

I apparently changed tools/picard/tool_data_table_conf.xml.test when writing Picard_CollectSequencingArtifactMetrics.xml, I think... 

What are you exactly suggesting now? Should I now open a new PR where I reset tools/picard/tool_data_table_conf.xml.test to its previous state at the start of this PR? 

@bgruening, @bernt-matthias and @nsoranzo: Thank you a lot!

**Comment by bgruening** [comment]
Just remove all your changes from tools/picard/.

You can do this for example: `git checkout tools/picard/`

**Comment by aerijman** [comment]
Am I finished with my duties? ;)

**Comment by bwlang** [comment]
This is looking good to me now - quite a slog!
@aerijman  thanks or working your way through it.  Hopefully you can take a few minutes to crystallize what you've learned about testing with cached data to improve the planemo documentation.

**Comment by bernt-matthias** [comment]
What do you mean by

> testing with cached data

**Comment by bwlang** [comment]
by 

> What do you mean by
> 
> > testing with cached data

Hi @bernt-matthias 
Maybe cached is not the right term...I'm referring to tests against the administrator-maintained data tables instead of those found in a history.

**Comment by bgruening** [comment]
@aerijman I have seen that you `	  <!--<param name="bed_filename" value="bedfile.bed"/>-->` comment out, but it seems this is an essential feature of the tool. Is that correct? Should it be out of the test?

**Comment by aerijman** [comment]
@bgruening Thank you for raising this question!
The parameter works with planes test. It's a conditional parameter (`optional="true"`).  
During the test, however (after the `if` statement?) `bed_filename` becomes `None`.  
Why is that?

File: tools/tasmanian_mismatch/test-data/cached_locally/all_fasta.loc [hunk 19]
```diff
@@ -0,0 +1,2 @@
+#chrM	chrM	ChrM	${__HERE__}/small_region.fa
```
Comment (bwlang) [inline tools/tasmanian_mismatch/test-data/cached_locally/all_fasta.loc:1]: remove

File: tools/tasmanian_mismatch/test-data/all_fasta.loc [hunk 20]
```diff
@@ -0,0 +1,2 @@
+#chrM	chrM	ChrM	${__HERE__}/cached_locally/picard_BedToIntervalList_ref.fa
```
Comment (bwlang) [inline tools/tasmanian_mismatch/test-data/all_fasta.loc:1]: remove

Comment (aerijman) [reply tools/tasmanian_mismatch/test-data/all_fasta.loc:1]: Removed. 
Also the conditional is resolved!

File: tools/tasmanian_mismatch/Tasmanian.xml [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+  </outputs>
+
+  <tests>
+    <!-- test when reference from history -->
+    <test>
+      <param name="bam_input" value="test2.bam" ftype="bam"/>
+      <param name="reference_source_selector" value="history"/>
+      <param name="ref_file" value="small_region.fa"/>
+
+	  <conditional name="bed_filename">
```
Comment (bgruening) [inline tools/tasmanian_mismatch/Tasmanian.xml:87]: Why is this here a conditional, its not above?

File: tools/tasmanian_mismatch/Tasmanian.xml [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+	  </conditional>
+
+      <output name="output_table" file="test2.output" lines_diff="4"/>
+    </test>
+    <!-- test when reference from cached-->
+    <test>
+      <param name="bam_input" value="test2.bam" ftype="bam" dbkey="hg38"/>
+      <param name="reference_source_selector" value="cached"/>
+      <param name="ref_file" value="hg38"/>
+	  <!--<param name="bed_filename" value="bedfile.bed"/>-->
```
Comment (bgruening) [inline tools/tasmanian_mismatch/Tasmanian.xml:98]: And this one? Remove or include?

File: tools/tasmanian_mismatch/Tasmanian.xml (hunk 21 above)
Comment (aerijman) [reply tools/tasmanian_mismatch/Tasmanian.xml:87]: It's the way I made the test work with a param with attribute `optional="true"`.   
It passes lint and test.   
Is it wrong?

File: tools/tasmanian_mismatch/Tasmanian.xml (hunk 22 above)
Comment (aerijman) [reply tools/tasmanian_mismatch/Tasmanian.xml]: removed it! I already include bed in the first test.

File: tools/tasmanian_mismatch/Tasmanian.xml (hunk 21 above)
Comment (bgruening) [reply tools/tasmanian_mismatch/Tasmanian.xml]: Yes, your bed_filename parameter is not included in any conditional, isn't it? So it needs to be a normal parameter in your test as well.

Comment (aerijman) [reply tools/tasmanian_mismatch/Tasmanian.xml]: In the [documentation](https://planemo.readthedocs.io/en/latest/writing_standalone.html#conditional-parameters),  `optional="true"` is listed under conditional parameters. Can be treated as such? Also, why doesn't it work in the test as a regular parameter? I get `None` instead of the __filename__. 
Thanks!

Comment (bgruening) [reply tools/tasmanian_mismatch/Tasmanian.xml]: @aerijman your commandline of the failing tests looks like `[ln -s '/tmp/tmp62xj4lff/files/c/e/1/dataset_ce117836-c955-444f-ba6a-1a78132ee2f1.dat' 'localref.fa' &&  samtools view '/tmp/tmp62xj4lff/files/b/8/7/dataset_b87eacc7-c826-4e00-9bed-ca78d762a7f5.dat' |  run_intersections -b '/tmp/tmp62xj4lff/files/f/c/7/dataset_fc79888a-1654-42e1-bca8-77a0e777b2d3.dat' |  run_tasmanian -q '20' -s '0' -m '20' -c '20' -r 'localref.fa' > '/tmp/tmp62xj4lff/files/c/8/a/dataset_c8a5c06a-a81c-4e64-9165-00f54baf12c2.dat'`

Which looks good to me. There is no None. None you only get when no file is specified. That why you have the `if $bedfile` in your command block.

The test is imho failing because of differences in the result file.

**Comment by bgruening** [comment]
Thanks a lot @aerijman! Sorry for being so nitpicking ...

**Comment by aerijman** [comment]
@bgruening: Actually, thank you a lot for being nitpicking!! It allows me to learn a lot and allows me to build a better tool-wrapper!

**Comment by bgruening** [comment]
Cool, glad it was helpful!
